src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -1279,12 +1279,12 @@ public SolutionState WithAnalyzerReferences(IReadOnlyList<AnalyzerReference> ana
             return null;
 
         // Do a quick check if the full info for that path has already been computed and cached.
-        var fileMap = _lazyFilePathToRelatedDocumentIds;
-        if (fileMap != null && fileMap.TryGetValue(filePath, out var relatedDocumentIds))
+        if (_lazyFilePathToRelatedDocumentIds.TryGetValue(filePath, out var relatedDocumentIds))
         {
             foreach (var relatedDocumentId in relatedDocumentIds)
             {
-                if (relatedDocumentId != documentId)
+                // Don't return documents from the same project
+                if (relatedDocumentId.ProjectId != documentId.ProjectId)
```

Comment:
>if (relatedDocumentId.ProjectId != documentId.ProjectId) [](http://example.com/codeflow?start=16&length=56)

This makes the cached hit logic follow the same as the non-cached logic from before. This is the expected behavior per Workspace.UpdateAddedDocumentToExistingContentsInSolution contract verification against the project ids.